bazel/PPROF.md

```diff
@@ -155,3 +175,145 @@ More complex flame/graph charts can be generated and viewed in a browser, which
 is often more helpful than text-based output:
 
     $ pprof -http=localhost:9999 bazel-bin/source/exe/envoy main_common_base*
+
+# Alternatives to `gperftools`
+
+## On-CPU analysis
+
+By default Envoy is built without gperftools. In this case the same results can be
+achieved for On-CPU analysis with the `perf` tool. For this there is no need to tweak
+Envoy's environment, you can even do measurements for an instance running in production
+(beware of possible performance hit though). Simply run:
+```
+$ perf record -g -F 99 -p `pgrep envoy`
+^C[ perf record: Woken up 1 times to write data ]
+[ perf record: Captured and wrote 0.694 MB perf.data (1532 samples) ]
+```
+
+The program will store the collected sampling data in the file `perf.data` whose
+format is also understood by recent enough versions of `pprof`:
+```
+$ pprof -http=localhost:9999 perf.data
+```
+## Memory analysis
+
+Unfortunately `perf` doesn't support heap profiling analogous to `gperftools`, but still
+we can get some insight into memory allocations with
+[Brendan Gregg's tools](http://www.brendangregg.com/FlameGraphs/memoryflamegraphs.html).
+You'll need to have [bcc](https://github.com/iovisor/bcc) installed in your system and a
+copy of [FlameGraph](https://github.com/brendangregg/FlameGraph):
+```
+$ git clone https://github.com/brendangregg/FlameGraph
+$ sudo /usr/share/bcc/tools/stackcount -p `pgrep envoy` \
+    -U "/full/path/to/envoy/bazel-bin/source/exe/envoy-static:_Znwm" > out.stack
+$ ./FlameGraph/stackcollapse.pl < out.stacks | ./FlameGraph/flamegraph.pl --color=mem \
+    --title="operator new(std::size_t) Flame Graph" --countname="calls" > out.svg
+```
+
+The `stackcount` utility counts function calls and their stack traces using eBPF probes.
+Since Envoy by default links statically to tcmalloc which provides its own implementation
+of memory management functions the used uprobe looks like
+```/full/path/to/envoy/bazel-bin/source/exe/envoy-static:_Znwm```. The part before
+the colon is a library name (a path to Envoy's binary in our case). The part after the
+colon is a function name as it looks like in the output of `objdump -tT /path/to/lib`,
+that is mangled in our case. To get an idea how your compiler mangles the name you
+can use this one-liner:
```

Comment:
can't [c++filt](https://sourceware.org/binutils/docs/binutils/c_002b_002bfilt.html) iron this all out?

Reply:
No, it can demangle only.